### ISSUE TITLE
[question] using release packages in a cmake debug build

### ISSUE THREAD

**Issue by MaxSagebaum** [conan-io/conan#17390]
### What is your question?

I am trying to evaluate conan as a package manger. What I usually do is to have different cmake configurations, e.g. `build_debug`, `build_release`, etc. I would like to always use the release packages from the conan repository.

I tried this use case on a small project. I installed everyting with:
```
conan install  . --output-folder=build_conan --build=missing
```

The release cmake configuration was generated with: (in `build_conan`)
```
cmake .. -DCMAKE_TOOLCHAIN_FILE=conan_toolchain.cmake -DCMAKE_BUILD_TYPE=Release
```

The debug cmake configuration was generated with: (in `build_conan_debug`)
```
cmake .. -DCMAKE_TOOLCHAIN_FILE=../build_conan/conan_toolchain.cmake -DCMAKE_BUILD_TYPE=Debug
```

The release version build without any problems. The Debug version could not find the header files.

I tried to read the documentation on https://docs.conan.io/2/tutorial/consuming_packages/different_configurations.html but this only describes how to have different conan configurations and not how to use the same conan configuration on different cmake configurations.

My `conanfile.txt`:
```
[requires]
lua/5.4.6
nlohmann_json/3.11.3
tclap/1.2.5

[generators]
CMakeDeps
CMakeToolchain
```

My CMakeLists.txt`:
```
cmake_minimum_required(VERSION 3.4)
project(ConanTest)

find_package(Lua REQUIRED)
find_package(TCLAP REQUIRED)

include_directories(include)

set(SOURCE_FILES src/conanTest.cpp)
add_executable(ConanTest ${SOURCE_FILES})

target_link_libraries(ConanTest tclap::tclap)
target_link_libraries(ConanTest lua::lua)

```

The documentation on https://docs.conan.io/2/tutorial/consuming_packages/the_flexibility_of_conanfile_py.html#consuming-packages-flexibility-of-conanfile-py-use-layout also seems to assume that the release and debug versions are installed.

Maybe I have missed a documentation. If so, I would be grateful for the link.

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @MaxSagebaum 

Thanks for your question.

By default, Conan manages all different binary configurations (os, compiler, arch, etc) as separate binaries, so it means that it is necessary to install both configurations:

```
conan install ... -s build_type=Release  # this is the almost always the default, not needed
conan install ... -s build_type=Debug 
```

Here the ``layout`` of the conanfile plays an important role. As you didn't define a ``[layout]`` in your ``conanfile.txt``, Conan might pollute your folder, or might not do the right thing by default if you don't manage it yourself. For example, for "single-config" systems (like gcc), something like this:

```
conan install ... -s build_type=Release  -of=conan_release
conan install ... -s build_type=Debug -of=conan_debug
```
while for multi-config systems, like MSVC, the same folder can be used for both:
```
conan install ... -s build_type=Release  -of=conan_both
conan install ... -s build_type=Debug -of=conan_both  
```
Where ``conan_both`` is just an example

But if you add the ``[layout]`` with ``cmake_layout`` to your ``conanfile.txt, then:
```
conan install ... -s build_type=Release 
conan install ... -s build_type=Debug 
```
will do the right thing, not polluting your folder, and selecting the right folder layout depending on the generator used.

Also, it would be recommended to use the generate presets, instead of picking the toolchain manually, do:
```
#msvc
cmake --preset conan-default 
# gcc
cmake --preset conan-release
```
And then to build, open the project or:
```
cmake --build --preset conan-release
```

Besides that, Conan has tons of configuration capabilities for the layout, like changing the base folder, or creating folders for different combinations, depending on arch, compiler, shared/static, etc, so if you are working simultaneously with multiple configurations, that can be convenient.

So in summary:
- It seems you were missing the ``conan install ... -s build_type=Debug `` to install the Debug configuration
- It is recommended to use the ``cmake_layout`` and the CMake presets

Please try that and let me know. Thanks for your feedback!

**Comment by MaxSagebaum** [comment]
Hi @memsharded,

thanks for the quick answer. 

I think the point is, that I do not want to do `conan install ... -s build_type=Debug `. Following cases why this is not always possible:
 - My dependencies are 40 GB of binary data. I do not want to have the same dependencies with debug data which would be like 120 GB of data.
 - My application is just a small plugin for a large application. I need to debug my plugin, but the main application needs to run with release binaries. E.g. time to break point with release binaries is 10 seconds, time to break point with debug binaries is 60 minutes.

By searching for `conan cmake layout` if found the tool documentation. (Google links to the 1.7 documentation and not to the 2.*, maybe this should be fixed.)

In https://docs.conan.io/2/reference/tools/cmake/cmakedeps.html I probably found the solution to my problem:
```
cmake .. -DCMAKE_BUILD_TYPE=Debug -DCMAKE_TOOLCHAIN_FILE=../build_conan/conan_toolchain.cmake -DCMAKE_MAP_IMPORTED_CONFIG_DEBUG=Release
```
With this command I have a debug build that uses conan release packages. Is this the way to go or is there a different option?

Is there a way to have an agnostic `conan install` that uses the release version of the packages, but can be picked up by any cmake configuration without `-DCMAKE_MAP_IMPORTED_CONFIG_DEBUG=Release`?

> It is recommended to use the cmake_layout and the CMake presets

I tried that and I found the _magic_ it does rather confusing. A `conan install` in e.g. `build_my_new_conan` will use `build` as the default directory and nothing in the `build_my_new_conan` directory is added. If I then do just a `cmake .. --preset conan-release` it still does not add anything to `build_my_new_conan` folder.

In my workflow I usually have different release and debug configurations that all need to exist next to each other. e.g. using different linear algebra packages and having tests/binaries for all of them.

**Comment by memsharded** [comment]
Understood, just to make sure, a couple of counter-cases (for other readers) to:

> My dependencies are 40 GB of binary data. I do not want to have the same dependencies with debug data which would be like 120 GB of data.
My application is just a small plugin for a large application. I need to debug my plugin, but the main application needs to run with release binaries. E.g. time to break point with release binaries is 10 seconds, time to break point with debug binaries is 60 minutes.

There are situations when using the Debug dependencies are totally necessary. For example Visual Studio MSVC typically links the runtimes differently for Debug/Release, so mixing dependencies built with one Release runtime with a consumer trying to build in Debug will easily fail with link errors. So it depends on the platform.

Regarding your solution, I think the easiest at this moment would be to do:
```
conan install ... -s build_type=Release -s "&:build_type=Debug"
```
That means: install everything (all depenencies) for Release, but my current consumer project (``&`` is a placeholder for "the current consumer") will be using Debug. 
This way, you won't have to install more heavy dependencies, and Conan will be aware of it and work, and the local folders with the layout will also be correctly named after debug/release current configuration irrespective of the dependencies.

Regarding the ``CMAKE_MAP_IMPORTED_CONFIG_DEBUG`` CMake imported configurations, it depends on the CMake version. Latest ones (I think >=3.28) defaulted the mapping, so it might not be necessary to explicitly define it. Still, there might be some edge cases that will be addressed better by a new ``CMakeDeps`` generator we are working on. 

So probably the above ``-s "&:build_type=Debug"`` is the recommended way to go, please give it a try and let us know.

**Comment by MaxSagebaum** [comment]
Ok thanks for the answer. Your suggestion works and I think with this I can accommodate my usual workflow.

The bootstrap would be:
```
mkdir conan_install
conan install .. -s build_type=Release -s "&:build_type=Release" -of .
conan install .. -s build_type=Release -s "&:build_type=Debug" -of .
# and for all other combinations requires -DCMAKE_BUILD_TYPE=???
```

Probably in addition with a `.conanrc` file that define the cache to be in the same folder. This way a `rm -rf conan_install` removes everything that conan has installed for this project. 

Afterwards a _normal_ cmake configuration can be done:
```
mkdir build_debug; cd build_debug;     cmake .. -DCMAKE_TOOLCHAIN_FILE=../conan_install/conan_toolchain.cmake -DCMAKE_BUILD_TYPE=Debug
mkdir build_release; cd build_release; cmake .. -DCMAKE_TOOLCHAIN_FILE=../conan_install/conan_toolchain.cmake -DCMAKE_BUILD_TYPE=Release
```

Thanks for the help. My question is now resolved.

**Comment by memsharded** [comment]
Some comments:

- The ``-of=.`` should be the default, not necessary
- The ``-DCMAKE_BUILD_TYPE`` is only for cmake, conan doesn't understand it
- The ``.conanrc`` might not be necessary in most cases. The Conan cache in the default location can save repeated storage and transfer time. 

The most important is that:
```
cmake .. -DCMAKE_TOOLCHAIN_FILE=../conan_install/conan_toolchain.cmake -DCMAKE_BUILD_TYPE=Debug
```
Doesn't look right for single-config setups. If you are doing single-config, like gcc, the ``cmake_layout`` or you need a separate folder, and it will contain a separate ``conan_toolchain.cmake``. So there is a different ``conan_toolchain.cmake`` for each Debug/Release configuration. Because the toolchain can have different flags for different configurations. In your case the second ``conan install`` is overwriting the first one, invalidating possible custom changes. For example, you could do:

```
conan install .... -c tools.build:cxxflags=<someflags> -s build_type=Release
conan install ..... -c tools.build:cxxflags=<otherdebugflags> -s build_type=Debug
```
If you don't have different toolchains for single-config, the otherdebugflags will overwrite the someflags.

If you use the ``cmake_layout`` conan will manage to do this ,and the ``conan_toolchain.cmake`` will have generator expressions to handle the different flags for multi-config configurations.

Recall the 2 different flows:

Single config (one configure step for each build, one folder for each configuration, 2 conan_toolchain.cmake files). Linux, gcc, etc:

```
# one folder
cmake .. -DCMAKE_BUILD_TYPE=xxxxx
cmake --build .
# other folder
cmake .. -DCMAKE_BUILD_TYPE=yyyy
cmake --build .
```

Multi-config (one configure step for all builds, one folder for the configure, 1 conan_toolchain.cmake file) MSVC projects:

```
# one folder
cmake .. 
# two builds from same configuration
cmake --build . --config Release
cmake --build . --config Debug
```

Your setup looks like single-config case, but you are using the same folder for the install and conan_toolchain.cmake, that is not correct.

**Comment by MaxSagebaum** [comment]
> The -of=. should be the default, not necessary

It seems to be in my case. With no `'[layout]` specification conan puts all files next to the conanfile.txt without the `-of=.`

> The -DCMAKE_BUILD_TYPE is only for cmake, conan doesn't understand it

I did not put this in a conan command. But since you mention that conan does not understand that why is conan linking its builds very strongly to the default cmake targets? You mentioned the windows use case but when you use system libraries, there is also not debug version available?

> The .conanrc might not be necessary in most cases. The Conan cache in the default location can save repeated storage and transfer time.

I understand that and usually this is a very good choice. But there is no `conan uninstall` - which should remove everything conan has installed for this consumer. Maybe with `conan remove -l list.txt`. Does conan create a list of all packages installed for a consumer? With every version ever installed? Having the local cache for the consumer solves this by just deleting the cache. 

> The most important is that:
> cmake .. -DCMAKE_TOOLCHAIN_FILE=../conan_install/conan_toolchain.cmake -DCMAKE_BUILD_TYPE=Debug
>Doesn't look right for single-config setups. If you are doing single-config, like gcc, the cmake_layout or you need a separate folder, and it will contain a separate conan_toolchain.cmake. So there is a different conan_toolchain.cmake for each Debug/Release configuration. Because the toolchain can have different flags for different configurations. In your case the second conan install is overwriting the first one, invalidating possible custom changes. For example, you could do:

I can understand that and in such a case I would do it like that. But in my default use case, I want all dependencies as releases and use the same release conan configuration for every cmake configuration I have.

So maybe the `-s "&:build_type=Debug" ` option is not the right choice for my use case and I should use `-DCMAKE_MAP_IMPORTED_CONFIG_DEBUG=Release`.

> If you use the cmake_layout conan will manage to do this ,and the conan_toolchain.cmake will have generator expressions to handle the different flags for multi-config configurations.

I tried the cmake_layout and in a default configuration it might work nicely. In the case you need multiple different debug or release builds at the same time, you can not simply do `mkdir build_own; cmake .. -DCMAKE_BUILD_TYPE=Debug <special configuration options>`. cmake will fail to get the dependencies installed under `<project_root>/build/Debug/generators`. So you have to specify `-DCMAKE_TOOLCHAIN_FILE=../build/Debug/generators/conan_toolchain.cmake` again. So I am doing it right away.

> Your setup looks like single-config case, but you are using the same folder for the install and conan_toolchain.cmake, that is not correct.

Why it is not correct to use a single conan install for multiple cmake configurations? I could have 6 different conan build configuration and 30 different cmake configurations. If I want to change 1 conan build configuration then I have to go to all cmake configurations that use that one and update the conan install and the cmake configuration. If I can do the layout of this myself and use one conan installation for multiple cmake configurations, the task of managing this is much simpler.

**Comment by memsharded** [comment]
> I can understand that and in such a case I would do it like that. But in my default use case, I want all dependencies as releases and use the same release conan configuration for every cmake configuration I have.

There are other things besides dependencies. The dependencies might be the same, but the compiler flags injected by Conan might be different in Release and Debug. As suggested above, the ``-c tools.build:cxxflags`` in Conan allows to inject C++ flags. They go directly to the ``conan_toolchain.cmake``, so if someone wants to have different flags in Release and Debug, it is mandatory to have separate ``conan_toolchain.cmake``. Having different ``conan_toolchain.cmake`` doesn't change at all the fact that in both cases you are linking with the Release dependencies, the dependencies are exactly the same in both cases and with both ``conan_tooclhain.cmake`` files.

> So maybe the -s "&:build_type=Debug"  option is not the right choice for my use case and I should use -DCMAKE_MAP_IMPORTED_CONFIG_DEBUG=Release.

Not really, I think the ``-s "&:build_type=Debug"`` is good. It is not connected to dependencies at all, what I am saying about the toolchain applies also for projects that don't have any dependency at all.


> I understand that and usually this is a very good choice. But there is no conan uninstall - which should remove everything conan has installed for this consumer. Maybe with conan remove -l list.txt. Does conan create a list of all packages installed for a consumer? With every version ever installed? Having the local cache for the consumer solves this by just deleting the cache.

There are several commands to remove files from the cache, like using the ``--lru`` arguments, to keep both the cache to minimal size while not affecting projects in your computer that rely on those dependencies. We haven't seen yet an issue where the size of cache is a problem, but having a separate cache for every project, if you have many projects and they have large dependencies will make things way more inefficient both in terms of slower downloads, more builds when binaries are missing and multiplying by N the storage in disk. 

I mean, this is totally fine, if you want to do it that way, good, the feature is there for this reason. Just trying to illustrate the tradeoffs, just in case 🙂 

I am not fully sure that "multi-config" vs "single-config" projects difference is clear. In all cases there are multiple configurations like Release and Debug. In all cases you can use the Release dependencies even if the consumer project is Release or Debug. "single-config" doesn't mean that you only have 1 config, but that every config that you have has its own ``conan_toolchain.cmake`` and its own ``cmake`` configure step.

Are you using Visual Studio at all? It seems that you are not, so you would be in a "single-config" setup.

> I tried the cmake_layout and in a default configuration it might work nicely. In the case you need multiple different debug or release builds at the same time, you can not simply do mkdir build_own; cmake .. -DCMAKE_BUILD_TYPE=Debug <special configuration options>. cmake will fail to get the dependencies installed under <project_root>/build/Debug/generators. So you have to specify -DCMAKE_TOOLCHAIN_FILE=../build/Debug/generators/conan_toolchain.cmake again. So I am doing it right away.

Have you tried the ``build_folder_vars`` conf? (check the available confs with ``conan config list <pattern>``, like ``conan config list cmake`` for example). This allows you to have different folders for different configurations. You can put that in your profile and forget about it.

Then I'll probably check the presets and use the generated presets, they are also very convenient.
There are users using the cmake_layout that are doing tenths of different local builds, with different configurations based on settings, options, etc, and the ``cmake_layout`` will generate the correct ``conan_toolchain.cmake`` in all cases. For "multi-config" projects, it will generate 1 ``conan_toolchain.cmake`` for the different Debug/Release configurations, if you have 20 differerent variants, there will be 10 projects with 10 ``conan_toolchain.cmake`` files, and 20 builds, while in "single-config" there will be 20 projects with 20 ``conan_toolchain.cmake`` files.

> Why it is not correct to use a single conan install for multiple cmake configurations? I could have 6 different conan build configuration and 30 different cmake configurations. If I want to change 1 conan build configuration then I have to go to all cmake configurations that use that one and update the conan install and the cmake configuration. If I can do the layout of this myself and use one conan installation for multiple cmake configurations, the task of managing this is much simpler.

Because Conan is doing both the dependencies and the toolchain. Regarding the dependencies, that is totally correct, you could have 1 installation of dependencies for all possible cmake configurations. But if you are using the ``CMakeToolchain``, that is not part of the dependencies, that is part of the CMake configuration, and that ties the ``conan install`` outputs to the CMake builds. 

If you want, you might drop the ``CMakeToolchain`` and do whatever you choose. You might need to define the CMAKE_PREFIX_PATH and other variables to locate the generated ``xxx-config.cmake`` files, and you must ensure that the Conan ``settings`` used to install the dependencies are compatible with your current build (for example ``conan_toolchain.cmake`` will define things like the MSVC runtime based on ``compiler.runtime`` for MSVC, the libstdc++ used for Linux based on ``compiler.libcxx`` or the C++ standard based on ``compiler.cppstd`` settings). This is typically one of the utilities of the ``CMakeToolchain``, but if you manage it, you can drop it. The other thing that ``CMakeToolchain`` does is generate the CMake preset files, but if you dont want to use them either, then you are good.

Please let me know if this clarifies a bit the issue and if you have any further questions.

**Comment by MaxSagebaum** [comment]
Thanks for the answer and your time, I really appreciate your effort to set my head straight. ;) Just two quick questions:

> Have you tried the build_folder_vars conf? (check the available confs with conan config list <pattern>, like conan config list cmake for example). This allows you to have different folders for different configurations. You can put that in your profile and forget about it.

From the documentation on https://docs.conan.io/1/reference/conanfile/tools/cmake/cmake_layout.html, I understand that `build_folder_vars` works on settings in the conan toolchain and not on settings in project?

> Then I'll probably check the presets and use the generated presets, they are also very convenient.
There are users using the cmake_layout that are doing tenths of different local builds, with different configurations based on settings, options, etc, and the cmake_layout will generate the correct conan_toolchain.cmake in all cases. For "multi-config" projects, it will generate 1 conan_toolchain.cmake for the different Debug/Release configurations, if you have 20 differerent variants, there will be 10 projects with 10 conan_toolchain.cmake files, and 20 builds, while in "single-config" there will be 20 projects with 20 conan_toolchain.cmake files.

Do you have an documentation/example/tutorial for this? What do you mean by `settings` (conan or project), by options (conan or project), etc.?

> Are you using Visual Studio at all? It seems that you are not, so you would be in a "single-config" setup.
No, I am working on linux with the command line.

**Comment by memsharded** [comment]
> Thanks for the answer and your time, I really appreciate your effort to set my head straight. ;) Just two quick questions:

Happy to help! :) Just want to make sure you don't hit any issues later down the road

> From the documentation on https://docs.conan.io/1/reference/conanfile/tools/cmake/cmake_layout.html, I understand that build_folder_vars works on settings in the conan toolchain and not on settings in project?

The ``build_folder_vars`` enter as an argument to the ``cmake_layout`` and to the ``CMakeToolchain`` in two ways:
- It can define the output folders that the ``cmake_layout`` will use. Then, it will change the folder in which the ``conan_toolchain.cmake`` files are located.
- It will also define different CMake presets names in the generated presets and connect those presets to the folders defined in ``cmake_layout``.

It is true that having the folders will also generate the ``xxx-config.cmake`` files for the dependencies, for each configuration. This isn't at issue at all, these files will be identical if you are using always the Release dependencies, and every configuration will pick the same dependencies. And those files are text files, so size or speed generating them is not an issue.

> Do you have an documentation/example/tutorial for this? What do you mean by settings (conan or project), by options (conan or project), etc.?

Uhm, that is a good point, it is true that I can only find https://docs.conan.io/2/examples/tools/cmake/cmake_toolchain/extend_own_cmake_presets.html, but this is not the basic case of setting ``build_folder_vars``. We have a bunch of tests in the test suite, not that challenging to read if you are interested, but probably it is worth adding some explicit example in the docs.

I would probably recommend to just try it, add in your command line or your profile something like:
```
-c tools.cmake.cmake_layout:build_folder_vars="['settings.compiler', 'settings.compiler.version']"
```
When you do the install, and see the effect. Then you can add ``options.shared``, or even a const, and check.

**Comment by MaxSagebaum** [comment]
From you answer, I think I got the principle of `build_folder_vars`. It allows me to link `conan` specific build options to the generated names for the build folder in cmake. If I have project/consumer specific options I need to add them as a cmake preset and use the nearest conan created preset. If I have 4 different compilation options where each can have 4 different values I would need 256 presets in my cmake file, when I want to use cmake_layout. Combining this with conan presets would multiply it with this number again. Maybe the python version of the `conanfile.txt` could be used to crate all these versions automatically. 

If I do not want have these cmake presets, then I am not able to use `cmake_layout` in conan since it forces me to use `build/Debug` and `build/Release` for all cmake specific settings.

I think the way with `-DCMAKE_TOOLCHAIN_FILE=..` and generated a release wrapper for all configurations covers my basic uses case. But I have to say it makes it quite involved to setup. It still bugs me, that conan does not allow me to generate a "general" conan install, that can be picked up by any cmake configuration out of the box.

__Something I noticed:__
You seem to really strongly advocating for the `cmake_layout` but in the tutorials (https://docs.conan.io/2/tutorial/consuming_packages.html) this is not explained. Maybe a tutorial on this should be added and an explanation about the constraints of this method.

__Something else:__
In https://docs.conan.io/2/tutorial/consuming_packages/intro_to_versioning.html#version-ranges you describe a few possibilities, how version ranges can be given. But one important part that is missing is how complete this description is. Are the examples everything? Is there a formal syntax definition on how to specify this? Such descriptions by example usually have the drawback that the reader can only use the examples. If his use case is not covered, he/she does not know if it is not possible or just not described. What is usually missing is a link to the full/formal description.

**Comment by memsharded** [comment]
> From you answer, I think I got the principle of build_folder_vars. It allows me to link conan specific build options to the generated names for the build folder in cmake. If I have project/consumer specific options I need to add them as a cmake preset and use the nearest conan created preset. If I have 4 different compilation options where each can have 4 different values I would need 256 presets in my cmake file, when I want to use cmake_layout. Combining this with conan presets would multiply it with this number again. Maybe the python version of the conanfile.txt could be used to crate all these versions automatically.

I am not sure if I follow. Nobody needs to create 256 different presets. Depending on the flow, Conan will generate the presets for you, so you only need to define the different configurations you want to install in different profiles, typically just a few, like 4-10 profiles to create up to 10 different binary configurations are common. Then you define in the layout how these profiles map to build folders/projects. In gcc-like projects if you want to be able to work with incremental builds in 10 different configurations, you want to map those configurations so they create 10 different folders.

So the flow is just:
```
conan install .. -pr=profile1
conan install .. -pr=profile2
...
cmake --preset conan-preset1  # name depends on build_folder_vars
...
cmake --preset conan-preset2
```
> But I have to say it makes it quite involved to setup

This is the part that I am also missing, why this would be involved to setup

> If I do not want have these cmake presets, then I am not able to use cmake_layout in conan since it forces me to use build/Debug and build/Release for all cmake specific settings.

Generated presets are a side product of ``CMakeToolchain`` that generates the ``conan_toolchain.cmake`` files. So the issue is not really associated to presets, even if not using presets, the toolchain files must be different. You cannot use the same toolchain file for all different configurations.

> I think the way with -DCMAKE_TOOLCHAIN_FILE=.. and generated a release wrapper for all configurations covers my basic uses case. But I have to say it makes it quite involved to setup. It still bugs me, that conan does not allow me to generate a "general" conan install, that can be picked up by any cmake configuration out of the box.

I might be missing something. Conan via ``CMakeDeps`` generator can perfectly generate an install that can be used by any consumer. (There are a couple of minor quirks that are being addressed in a new CMakeDeps generator, see https://docs.conan.io/2/incubating.html, but that is a different story, shouldn't be a blocker for this). 

The issue is if using ``CMakeToolchain``, because its purpose is effectively to be different for each configuration. Using ``options shared=True`` generates a toolchain with ``set(BUILD_SHARED_LIBS ON)``, and using a setting like ``compiler.runtime=static`` in Windows will generate a toolchain with ``set(CMAKE_MSVC_RUNTIME_LIBRARY Multithreaded)``. That is what toolchains are for. Maybe you just want to drop the ``CMakeToolchain`` generator?


Maybe we could communicate better via code? Like if we have a toy project that emulates what you are trying to achieve, and a script that illustrate the flows/setup, then we could collaborate via code.

**Comment by MaxSagebaum** [comment]
The toy project is a good idea. I created one with https://github.com/MaxSagebaum/conan_example

`bootstrap_conan.sh` creates the conan configurations.
`bootstrap.sh` creates the application configurations.

 - It uses as a dependency mpi which I definitely  do not want to have as debug build.
 - I added two options which have each 4 values.
 - `FUNCTION_TYPE` could switch between some library, different implementations of an algorithm, etc. 
 - `VALUE_TYPE` could switch between some computational type. (I am doing algorithmic differentiation (AD), here we have eg. `codi::RealReverse`, `codi::RealReverseIndex` as different drop in replacements for double.

What would be nice:
 - Have only one conan install. 
 - Have conan install all dependencies in a given root folder. (I did not check if this is possible.)
 - Remove the need to specify the `-DCMAKE_TOOLCHAIN_FILE`.
 - Have proper errors in the cmake configuration if the wrong toolchain file is specified.

__Some explanations for the above points:__

For testing I usually want to configure all possible configuration. Normally. just in the release version but if something goes wrong, I need the debug version of the application (not the dependencies e.g. MPI). 

This needs to be usable by other users. So I need to explain to them, that they need to install the dependencies via conan. That is ok and can be one command. That they need to do this for each cmake build type will raise already some questions. "What have my dependencies to do with the build type of the application." If they do not do this properly, conan will  install the release and debug dependencies of MPI.

The second step I need to teach is that they have to use the argument `-DCMAKE_TOOLCHAIN_FILE=../conan_deps/Release/conan_toolchain.cmake`, so that cmake can pick up the dependencies. If they do not match the correct toolchain file with the cmake configuration, cmake will configure but they will get wired build errors. 

The two above abstracts is what I mean with `it makes the setup quite involved`.

One additional problem is the location of the MPI installation. In order to run the application you need to know where conan has installed MPI. For me this was `/home/msagebaum/.conan2/p/b/openm5c06c1ea72bc5/p/bin/mpirun`. Having  a install root directory with `bin`, `lib` and `include` would be nice. This would also allow cmake to pick up the libraries with environment variables.

**Comment by memsharded** [comment]
This ticket should now be fixed with the new CMakeConfigDeps generator: https://docs.conan.io/2/incubating.html, available in Conan 2.12. Please open a new ticket for any further issue with the CMakeConfigDeps generator if necessary.

I am closing this ticket original post with it, but I think there are other conversations, and there was even a PR in https://github.com/MaxSagebaum/conan_example/pull/1. It would be better to start a new ticket for any further discussion, questions or comments, as this one deviated a bit too much from the original post (which was addressed in responses above, and now the new ``CMakeConfigDeps`` can address even better)

To get the new CMakeConfigDeps out of "incubating", please test it and report your positive feedback in https://github.com/conan-io/conan/issues/18211 (for issues/questions, better open new tickets)


Many thanks!
